### PR TITLE
update docs related to partition table dynamic mode

### PR DESCRIPTION
**Pull request by qw4990**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by rebelice — APPROVED** [review]
LGTM

**Review by lysu — APPROVED** [review]

**Comment by lysu** [comment]
please fix lint~

File: experimental-features.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ aliases: ['/docs-cn/dev/experimental-features-4.0/','/zh/tidb/dev/experimental-f
 
 + List 分区表 （List Partition）。（v5.0 实验特性）
 + List Column 分区表 （List Column Partition)。（v5.0 实验特性）
++ [分区表动态模式](/partitioned-table.md#动态模式)。
```
Comment (qiancai) [inline experimental-features.md:27]:
```suggestion
+ [分区表动态模式](/partitioned-table.md#动态模式)。（v5.1 实验特性）
```

**Review by tiancaiamao — APPROVED** [review]

**Comment by qw4990** [comment]
@lysu updated. Please approve it again~

**Review by lysu — APPROVED** [review]

File: partitioned-table.md [hunk 2]
```diff
@@ -1240,3 +1240,173 @@ select * from t;
...
 
 该变量仅作用于建表，已经建表之后再修改该变量无效。详见[系统变量和语法](/system-variables.md#tidb_enable_list_partition-从-v50-版本开始引入)。
+
+### 动态模式
+
+> **警告：**
+>
+> 该功能目前为实验特性，不建议在生产环境中使用。
+
+TiDB 访问分区表有两种模式，`dynamic` 和 `static`，目前默认情况下使用 `static` 模式，如果想打开动态模式，需要手动将 `tidb_partition_prune_mode` 设置为 `dynamic`。
```
Comment (TomShawn) [inline partitioned-table.md:1250]:
```suggestion
TiDB 访问分区表有两种模式，`dynamic` 和 `static`，目前默认使用 `static` 模式。如果想开启 `dynamic` 模式，需要手动将 `tidb_partition_prune_mode` 设置为 `dynamic`。
```

File: partitioned-table.md [hunk 3]
```diff
@@ -1240,3 +1240,173 @@ select * from t;
...
+
+{{< copyable "sql" >}}
+
+```sql
+set @@session.tidb_partition_prune_mode = 'dynamic'
+```
+
+在 `static` 模式下，TiDB 用多个算子单独访问每个分区，然后通过 Union 将结果合并起来；而在 `dynamic` 模式下，单个算子直接访问多个分区，不再使用 Union。
```
Comment (TomShawn) [inline partitioned-table.md:1258]:
```suggestion
在 `static` 模式下，TiDB 用多个算子单独访问每个分区，然后通过 Union 将结果合并起来。在 `dynamic` 模式下，单个算子直接访问多个分区，不使用 Union。
```

File: partitioned-table.md [hunk 4]
```diff
@@ -1240,3 +1240,173 @@ select * from t;
...
+
```
Comment (TomShawn) [inline partitioned-table.md:1259]: 需要概述一下这个示例做了哪些操作。

File: partitioned-table.md [hunk 5]
```diff
@@ -1240,3 +1240,173 @@ select * from t;
...
+| id                      | estRows  | task      | access object   | operator info                  |
++-------------------------+----------+-----------+-----------------+--------------------------------+
+| TableReader_7           | 3323.33  | root      | partition:p0,p1 | data:Selection_6               |
+| └─Selection_6           | 3323.33  | cop[tikv] |                 | lt(test.t1.id, 150)            |
+|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t1        | keep order:false, stats:pseudo |
++-------------------------+----------+-----------+-----------------+--------------------------------+
+3 rows in set (0.00 sec)
+```
+
+从上面例子可以看出，打开动态模式后，执行计划便没有了 Union 算子，且分区裁剪依然生效，都只范围了 `p0` 和 `p1` 两个分区。
```
Comment (TomShawn) [inline partitioned-table.md:1298]: \`dynamic\` 模式或动态模式，需要在文档中保持一致。
```suggestion
从以上示例可见，开启 `dynamic` 模式后，执行计划便没有了 Union 算子，且分区裁剪依然生效，都只访问了 `p0` 和 `p1` 两个分区。
```

File: partitioned-table.md [hunk 6]
```diff
@@ -1240,3 +1240,173 @@ select * from t;
...
+
+`dynamic` 模式让执行计划变得更简单清晰，剔除 Union 操作提高执行效率，以及避免了 Union 并发管理的问题，此外还解决了两个 `static` 模式无法解决的问题：
```
Comment (TomShawn) [inline partitioned-table.md:1300]:
```suggestion
`dynamic` 模式让执行计划更简单清晰，省略 Union 操作可提高执行效率，还可避免 Union 并发管理的问题。此外 `dynamic` 模式还解决了两个 `static` 模式无法解决的问题：
```

File: partitioned-table.md [hunk 7]
```diff
@@ -1240,3 +1240,173 @@ select * from t;
...
+
+1. 不能够使用 Plan Cache；
+
+2. 不能够使用 IndexJoin 的执行方式；
```
Comment (TomShawn) [inline partitioned-table.md:1304]:
```suggestion
+ 不能使用 Plan Cache

+ 不能使用 IndexJoin 的执行方式
```

File: system-variables.md [hunk 8]
```diff
@@ -566,6 +566,16 @@ Query OK, 0 rows affected (0.09 sec)
 - 默认值：OFF
 - 这个变量用来设置是否开启 `LIST (COLUMNS) TABLE PARTITION` 特性。
 
+### `tidb_partition_prune_mode` <span class="version-mark">从 v5.1 版本开始引入</span>
+
+> **警告：**
+>
+> 目前 Partition Table Dynamic Mode 为实验特性，不建议在生产环境中使用。
```
Comment (TomShawn) [inline system-variables.md:573]:
```suggestion
> 目前分区表动态模式为实验特性，不建议在生产环境中使用。
```

File: system-variables.md [hunk 9]
```diff
@@ -566,6 +566,16 @@ Query OK, 0 rows affected (0.09 sec)
...
+
+- 作用域：SESSION | GLOBAL
+- 默认值：static
+- 这个变量用来设置是否开启动态模式，关于动态模式的详情请参阅[分区表动态模式](/partitioned-table.md#动态模式)。
```
Comment (TomShawn) [inline system-variables.md:577]:
```suggestion
- 这个变量用来设置是否开启分区表动态模式。关于动态模式的详细说明请参阅[分区表动态模式](/partitioned-table.md#动态模式)。
```

File: partitioned-table.md [hunk 10]
```diff
@@ -1240,3 +1240,173 @@ select * from t;
...
+
+下面是关于 Plan Cache 的例子：
+
+{{< copyable "sql" >}}
+
+```sql
+-- 需要提前在配置文件中代开 Plan Cache 这个功能
```
Comment (TomShawn) [inline partitioned-table.md:1311]:
```suggestion
-- 需要提前在配置文件中开启 Plan Cache 功能
```

File: partitioned-table.md [hunk 11]
```diff
@@ -1240,3 +1240,173 @@ select * from t;
...
+
+mysql> explain select * from t1 where id < 150;
++------------------------------+----------+-----------+------------------------+--------------------------------+
+| id                           | estRows  | task      | access object          | operator info                  |
++------------------------------+----------+-----------+------------------------+--------------------------------+
+| PartitionUnion_9             | 6646.67  | root      |                        |                                |
+| ├─TableReader_12             | 3323.33  | root      |                        | data:Selection_11              |
+| │ └─Selection_11             | 3323.33  | cop[tikv] |                        | lt(test.t1.id, 150)            |
+| │   └─TableFullScan_10       | 10000.00 | cop[tikv] | table:t1, partition:p0 | keep order:false, stats:pseudo |
+| └─TableReader_18             | 3323.33  | root      |                        | data:Selection_17              |
```
Comment (TomShawn) [inline partitioned-table.md:1278]: + 这两个代码块太大了，文档叙述主体拿到示例代码块中有些影响阅读体验。能否将代码块分隔开，尤其是将注释 -- 拿出来呢？
+ 注释也显得有些简略和随意，建议向读者清楚交代整个示例中，每一个关键操作的语句做了什么，结果是什么，在用户文档中才会显得正式和易懂。

**Comment by qw4990** [comment]
已更新，辛苦再 review 一下 @TomShawn

File: partitioned-table.md [hunk 12]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+| id                      | estRows  | task      | access object   | operator info                  |
++-------------------------+----------+-----------+-----------------+--------------------------------+
+| TableReader_7           | 3323.33  | root      | partition:p0,p1 | data:Selection_6               |
+| └─Selection_6           | 3323.33  | cop[tikv] |                 | lt(test.t1.id, 150)            |
+|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t1        | keep order:false, stats:pseudo |
++-------------------------+----------+-----------+-----------------+--------------------------------+
+3 rows in set (0.00 sec)
+```
+
+可见执行计划中的 Union 消失了，此外分区裁剪依然生效，上述执行计划只访问了 `p0` 和 `p1` 两个分区。
```
Comment (TomShawn) [inline partitioned-table.md:1304]:
```suggestion
从以上查询结果可知，执行计划中的 Union 消失了，分区裁剪依然生效，且执行计划只访问了 `p0` 和 `p1` 两个分区。
```

File: partitioned-table.md [hunk 13]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+
+`dynamic` 模式让执行计划更简单清晰，省略 Union 操作可提高执行效率，还可避免 Union 并发管理的问题。此外 `dynamic` 模式还解决了两个 `static` 模式无法解决的问题：
+
++ 不能使用 Plan Cache
+
++ 不能使用 IndexJoin 的执行方式
+
+下面是关于 Plan Cache 的例子，我们在配置文件中开启 Plan Cache 功能后，在 `static` 模式下执行同一个查询两次：
```
Comment (TomShawn) [inline partitioned-table.md:1312]:
```suggestion
**示例一**：以下示例在配置文件中开启 Plan Cache 功能，并在 `static` 模式下执行同一个查询两次：
```

File: partitioned-table.md [hunk 14]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+mysql> select @@last_plan_from_cache;
++------------------------+
+| @@last_plan_from_cache |
++------------------------+
+|                      0 |
++------------------------+
+1 row in set (0.00 sec)
+```
+
+`last_plan_from_cache` 变量可以指示上一次查询是否命中 Plan Cache，可见在 `static` 模式下，即使执行同一个分区表查询多次，也不会命中 Plan Cache。
```
Comment (TomShawn) [inline partitioned-table.md:1342]:
```suggestion
`last_plan_from_cache` 变量可以显示上一次查询是否命中 Plan Cache。从以上示例一可知，在 `static` 模式下，即使在同一个分区表上执行多次查询，也不会命中 Plan Cache。
```

File: partitioned-table.md [hunk 15]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+3 rows in set (0.00 sec)
+```
+
+可见执行计划中的 Union 消失了，此外分区裁剪依然生效，上述执行计划只访问了 `p0` 和 `p1` 两个分区。
+
+`dynamic` 模式让执行计划更简单清晰，省略 Union 操作可提高执行效率，还可避免 Union 并发管理的问题。此外 `dynamic` 模式还解决了两个 `static` 模式无法解决的问题：
+
++ 不能使用 Plan Cache
+
++ 不能使用 IndexJoin 的执行方式
```
Comment (TomShawn) [inline partitioned-table.md:1310]:
```suggestion
+ 不能使用 Plan Cache（见以下示例一和示例二）
+ 不能使用 IndexJoin 的执行方式（见以下示例三和示例四）
```

File: partitioned-table.md [hunk 16]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+| @@last_plan_from_cache |
++------------------------+
+|                      0 |
++------------------------+
+1 row in set (0.00 sec)
+```
+
+`last_plan_from_cache` 变量可以指示上一次查询是否命中 Plan Cache，可见在 `static` 模式下，即使执行同一个分区表查询多次，也不会命中 Plan Cache。
+
+下面是在 `dynamic` 模式下执行同样操作的情况：
```
Comment (TomShawn) [inline partitioned-table.md:1344]:
```suggestion
**示例二**：以下示例在 `dynamic` 模式下执行与示例一相同的操作：
```

File: partitioned-table.md [hunk 17]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+mysql> select @@last_plan_from_cache;
++------------------------+
+| @@last_plan_from_cache |
++------------------------+
+|                      1 |
++------------------------+
+1 row in set (0.00 sec)
+```
+
+可见打开 `dynamic` 模式后，分区表查询便能命中 Plan Cache 了。
```
Comment (TomShawn) [inline partitioned-table.md:1371]:
```suggestion
由示例二结果可知，开启 `dynamic` 模式后，分区表查询能命中 Plan Cache 。
```

File: partitioned-table.md [hunk 18]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+
+下面是关于 IndexJoin 的例子，同样先在 `static` 模式下，尝试执行带 IndexJoin 的执行计划：
```
Comment (TomShawn) [inline partitioned-table.md:1373]:
```suggestion
**示例三**：以下示例在 `static` 模式下尝试执行计划带 IndexJoin 的查询：
```

File: partitioned-table.md [hunk 19]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+|   │ └─Selection_29             | 9990.00  | cop[tikv] |                        | not(isnull(test.t1.id))                        |
+|   │   └─TableFullScan_28       | 10000.00 | cop[tikv] | table:t1, partition:p2 | keep order:false, stats:pseudo                 |
+|   └─TableReader_36             | 9990.00  | root      |                        | data:Selection_35                              |
+|     └─Selection_35             | 9990.00  | cop[tikv] |                        | not(isnull(test.t1.id))                        |
+|       └─TableFullScan_34       | 10000.00 | cop[tikv] | table:t1, partition:p3 | keep order:false, stats:pseudo                 |
++--------------------------------+----------+-----------+------------------------+------------------------------------------------+
+17 rows in set, 1 warning (0.00 sec)
+```
+
+上面例子可见，即使使用了 `TIDB_INLJ` 的 hint，也无法使得带分区表的查询选上 IndexJoin。
```
Comment (TomShawn) [inline partitioned-table.md:1409]:
```suggestion
从以上示例三结果可知，即使使用了 `TIDB_INLJ` 的 hint，也无法使得带分区表的查询选上带 IndexJoin 的执行计划。
```

File: partitioned-table.md [hunk 20]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+
+接下来是在 `dynamic` 模式下的情况：
```
Comment (TomShawn) [inline partitioned-table.md:1411]:
```suggestion
**示例四**：以下示例在 `dynamic` 模式下尝试执行计划带 IndexJoin 的查询：
```

File: partitioned-table.md [hunk 21]
```diff
@@ -1240,3 +1240,196 @@ select * from t;
...
+| └─IndexLookUp_10(Probe)         | 1.25     | root      | partition:all          |                                                                                                                     |
+|   ├─Selection_9(Build)          | 1.25     | cop[tikv] |                        | not(isnull(test.t1.id))                                                                                             |
+|   │ └─IndexRangeScan_7          | 1.25     | cop[tikv] | table:t1, index:id(id) | range: decided by [eq(test.t1.id, test.t2.id)], keep order:false, stats:pseudo                                      |
+|   └─TableRowIDScan_8(Probe)     | 1.25     | cop[tikv] | table:t1               | keep order:false, stats:pseudo                                                                                      |
++---------------------------------+----------+-----------+------------------------+---------------------------------------------------------------------------------------------------------------------+
+8 rows in set (0.00 sec)
+```
+
+可见打开 `dynamic` 模式后，IndexJoin 便能被选上了。
```
Comment (TomShawn) [inline partitioned-table.md:1435]:
```suggestion
从示例四结果可知，开启 `dynamic` 模式后，带 IndexJoin 的计划在执行查询时被选上。
```

**Review by TomShawn — APPROVED** [review]
LGTM

**Pull request merged**